### Chapter 10 Files and Exceptions
_In this chapter you'll learn to work with files so your programs can quickly analyze lots of data_ <br>
<br>
You'll learn about _exceptons_, which are special objects, Python creates to manage errors that arise while a program is running. You'll also <br>
learn about the json module, which allows you to save user data so it isn't lost when your program stops running.<br>
<br>
Learning to work with files and save data will make your programs easier for people to use.<br>

*** 
### Reading from a File
To begin, we need a file with a few lines of text in it. Let's start with a file that contains _pi_ to 30 decimal places with 10 decimal places <br>
per line:

In [2]:
#pi_digits.txt
3.1415926535
8979323846
2643383279

2643383279

Here's a program that opens this file, reads it, and prints the contents of the file to the screen:

In [1]:
with open('pi_digits.txt') as file_object:
    contents = file_object.read()
    print(contents)

To do any wortk with a file, even just printing its contents, you first need to _open_ the file to access it. The __open( )__ function needs one<br> argument: the name of the file you want to open. Python looks for this file in the directory where the program that's currently being exeuted is<br> stored. The __open( )__ function returns an object representing the file. <br>
<br>
The keyword __with__ closes the file once access to it is no longer needed. Notice How we call __open( )__ in this program but not __close( )__.<br>
You could open and close the file by calling __oepn( )__ and __close( )__, but if a bug in your program prevents the __close( )__ statement from <br>
being executed, the file may never close. This may seem trivial, but improperly closed files can cause data to be lost or corrupted. And if you <br>
call __close( )__ state too early in your program, you'll find yourself trying to work with a _closed_ file (a file you can't access), which leads to more errors. <br>
<br>
The blank line appears because __read( )__ returns an empty string when it reaches the end of the file; this empty string shows up as blank line.<br>
If you want to remove the extra blank line, you can use __rstrip( )__ in the __print__ statement.

In [ ]:
with open('pi_digits.txt') as file_object:
    contents = file_object.read()
    print(contents.rstrip())

Recall that Python's __rstrip( )__ method removes, or strips, any whitespace characters from the right side of a string. <br>
<br>
#### -_File Paths_
When you pass a simple filename to the __open( )__ function, Python looks in the directory where the file that's currently being executed <br>
(your _.py_ program file) is stored. <br>
<br>
Sometimes, the file you want to open won't be in the same directory as your program file. To get python to open files froma directory other than <br>
the one where your program file is stored, you need to provide a _file path_, which tells Python to look into a specific location in your system. <br>
<br>
A relative file path tells Python to look for a given location relative to the directory where the currently running program file is stored. For example, you'd write:

In [4]:
with open('Resources/pi_digits.txt') as file_object:

3.1415926535
  8979323846
  2643383279


This line tells PYthon to look for the desired _.txt_ file in the folder _Resources_. <br>
<br>
You can also tell Python exactly where the file is on your computer regardless of where the program that's being executed is stored. This is <br>
called an _absolute file path_:

In [25]:
file_path = 'C:/Users/Alexs/OneDrive/Desktop/study_material/Python_Study/Resources/pi_digits.txt'
with open(file_path) as file_object:

IndentationError: expected an indented block (Temp/ipykernel_18940/2794869032.py, line 2)

#### -_Reading Line by Line_
When you're reading a file, you'll often want to examine each line of the file. You might be looking for a certain information in the file, or you<br>
might want to modify the text in the file in some way. For example, you might want to read through a file of weather data and work with any line <br>
that includes the word _sunny_ in the description of that day's weather. <br>
<br>
You can use a __for__ loop on the file object to examine each line from a file one at a time:

In [28]:
file_path= 'Resources/pi_digits.txt'

with open(file_path) as file_object:
    for line in file_object:
        print(line)

3.1415926535

  8979323846

  2643383279



These blank lines appear because an invisible newline character is at the end of each line in the text file. The __print__ statement adds its own <br>
newline each time we call it, so we end up with two newline characters at the end of each line: one from the file and one from the __print__ <br>
statement. Using __rstrip( )__ on each line in the __print__ statement eliminates these extra blank lines:

In [30]:
file_path= 'Resources/pi_digits.txt'

with open(file_path) as file_object:
    for line in file_object:
        print(line.rstrip())

3.1415926535
  8979323846
  2643383279


#### -_Making a List of Lines from a File_
When you use __with__, the file object returned by __open( )__ is only available inside the __with__ block that contains it. If you want to retain<br>
access to a file's content outside the __with__ block, you can store the file's lines in a list inside the block and then work with that list. <br>
You can process parts of the file immediately and postpone some processing for later in the program. <br>
<br>
The following example stores the lines of _pi_digits.txt_ in a list inside the __with__ block and the prints the lines outside the __with__ block:

In [2]:
file_path = 'Resources/pi_digits.txt'

with open(file_path) as file_object:
    lines = file_object.readlines()

for line in lines:
    print(line.strip())

3.1415926535
8979323846
2643383279


The __readlines()__ method takes each line from the file and stores it in a list. This list is then stored in __lines__, which we can continue <br>
to work with after the __with__ block ends. We use a simple __for__ loop to print each line from __lines__. <br>
<br>
#### -_Working with a File's Contents_
After you've read a file into memory, you can do whatever you want with that data, so let's briefly explore the digits of _pi_. First, we'll <br>
attempt to build a single string containing all the digits in the file with no whitespace in it:

In [1]:
filename = 'Resources/pi_digits.txt'

with open(filename) as file_object:
    lines = file_object.readlines()

pi_string = " "

for line in lines:
    pi_string += line.rstrip()

print(pi_string)
print(len(pi_string))

 3.1415926535  8979323846  2643383279
37


The variable pi_string contains the whitespace that was on the left side of the digits in each line, but we can get rid of that by using <br>
__strip( )__ instead of __rstrip( )__:

In [3]:
file_path = "Resources/pi_digits.txt"
with open(filename) as file_object:
    lines = file_object.readlines()

pi_string = " "
for line in lines:
    pi_string += line.strip()

print(pi_string)
print(len(pi_string))

 3.141592653589793238462643383279
33


#### -_Large Files: One Million Digits_
So far we've focused on analyzing a text file that contains only three lines, but the code in these examples would just as well on much larger <br> files. Python has no inherent limit to how much data you can work with. <br>
#### -_Is Your Birthday Contain in Pi_
Let's use the program we just wrote to find out if someone's birthday appears anywhere in the first million digits of _pi_. We can do this by <br> 
expressing each birthday as a string of digits and seeing if that string appears anywhere in pi_string:

In [5]:
file_path = "Resources/pi_million_digits.txt"
with open(file_path) as file_object:
    lines = file_object.readlines()
    
pi_string = " "
for line in lines:
    pi_string += line.strip()

birthday = input("Enter your birthday, in the form mmddyy: ")
if birthday in pi_string:
    print("Your birthday appears in the first million digits of pi!")

else:
    print("Your birthday does not appear in the first million digits of pi.")

Enter your birthday, in the form mmddyy:  010899


Your birthday appears in the first million digits of pi!


***
### Writing to a File
One of the simplest ways to save data is to write it to a file. When you write text to a file, the output will still be available after you <br>
close the terminal containing your program's output. You can exmine output after a program finishes running, and you can share the output files <br>
with others as well. You can also write programs that read and text back into memory and work with it again later. <br>
#### -_Writing an Empty File_
To write text to file, you need to call __open( )__ with a second argument telling Python that you want to write to the file. To see how this<br>
works let's write a simple message and store it in a file instead of priting it to the screen:

In [8]:
file_name = 'Resources/programming.txt.'
with open(file_name, 'w') as file_object:
    file_object.write("I love programming.")

The call to __open( )__ in this example has two arguments. The first argument is still the name of the file we want to open. The second argument <br>
'w' tells Python that we want to open the file in _write mode_. You can open a file in _read mode_ ('r'), _write mode_ ('w'), _append mode_ ('a')<br>
or a mode that allows you to read and write to the file ('r+'). If you omit the mode argument Python opens the file in read-only mode by default. <br>
<br>
The __open( )__ function automatically creates the file you're writing to if it doesn't already exist. However, be careful opening a file in the <br>
write mode ('w') because if the file does exist, Python will erase the file before returning the file object. <br>
<br>
We use the __write( )__ method on the file object to write a string to the file. <br>
<br>
>Note Python can only write strings to a text file. If you want to store numerical data in a text file, you'll have to convert the data to string<br>
format first using the str( ) function. <br>
<br>
#### -_Writing Multiple Lines_
The __write( )__ function doesn't add any newlines to the text you write. So if you write more than one line without including newlines <br>
characters, your file may not look the way you want it to:

In [16]:
filename = 'Resources/programming.txt'
with open(filename, 'w') as file_object:
    file_object.write("I love programming.")
    file_object.write("\nI love creating new games.")

If we open _programming.txt_, you'll see the two lines squished together. <br>
> I love programming.I love creating new games. <br>
<br>

Including the newlines in your __write( )__ statenents makes each string appear on its own line:
> I love programming. <br>
> I love creating new games. <br>
<br>

You can also use spaces, tab characters, and blank lines to format your output, just as you've been doing with the terminal-based output. <br>
<br>
#### -_Appending to a File_
If you want to add content to a file instead of writing over existing content, you can open the file in _append mode_. When you open a file in append mode, Python doesn't erase the file before <br>
returning the file object. Any lines you write to the file will be added to the end of the file. If the file doesn't exist yet, Python will create an empty file for you. 

In [4]:
filename = 'Resources/programming.txt'
with open(filename, 'a') as file_object:
    file_object.write("I also love finding meaning in large datasets.\n")
    file_object.write("I love creating apps that can run in a browser.\n")

We just end up with the original contents of the file, followed by the new content we just added. <br>
<br>
### Exception
Python uses special objects called _exceptions_ to manage errors that arise during a programs execution. Exceptions are handled with __try-except__ blocks. A __try-except__ block asks Python to do <br>
something, but it also tells Python what to do if an exception is raised. <br>
#### -_Using try-except Blocks_
When you think an error may occur, you can write a __try-except__ block to handle the exception that might be raised. You tell Python to try running some code, and you tell it what to do if the code <br>
results in a particular kind of exception. <br>
<br>
Here's what a __try-except__ block looks like:

In [1]:
try:
    print(5/0)
except ZeroDivisionError:
    print("You can't divide by zero!")

You can't divide by zero!


If the code in a try block works, Python skips over the except block. If the code in the try block causes an error, Python looks for an except block whose error matches the one that was raised <br>
and runs the code in that block. <br>
#### - _Using Exceptions to Prevent Crashes_
Handling errors correctly is especially important when the program has more work to do after the error occurs. This happens often in programs that prompt users for input. If the program responds <br>
to invalid input appropriately, it can prompt for move valid input instead of crashing. 

In [7]:
print("Give me two numbers, and I'll divide them.")
print("Enter 'q' to quit.")

while True:
    first_number = input("\nFirst number: ")
    if first_number == 'q':
        break
    second_number = input("\nSecond number: ")
    if second_number == 'q':
        break
    answer = int(first_number) / int(second_number)
    print(answer)

Give me two numbers, and I'll divide them.
Enter 'q' to quit.



First number:  5

Second number:  4


1.25



First number:  q


This program does nothing to handle errors, so asking it to divide by zero causes it to crash. <br>
#### -_The else Block_
We can make this program more error resistant by wrappingthe line that might produce errors in a __try-except__ block. The error occurs on the line that performs the division, so that's where we'll <br> put the __try-except__ block. 

In [ ]:
print("Give me two numbers, and I'll divide them.")
print("Enter 'q' to quit.")

while True:
    first_number = input("\nFirst number: ")
    if first_number == 'q':
        break
    second_number = input("\nSecond number: ")
    try:
        answer = int(first_number) / int(second_number)
    except ZeroDivisionError:
        print("You can't divide by 0!")
    else:
        print(answer)

#### -_Handling the FileNotFoundError Exception_
One common issue when working with files is handling missing files. The file you're looking for might be in a different location, the filename may be misspelled, or the file may not exist at all.<br>
You can handle all of these situations in a straightforward way with a __try-except__ block. Let's read a file that doesn't exist. The following program tries to read in the contents of <br>
_Alice in Wonderland_, but it isn't saved the file in same directory:

In [12]:
filename = 'alice.txt'
try:
    with open(filename, encoding = 'utf-8') as f_obj:
        contents = f_obj.read()
except FileNotFoundError:
    msg = "Sorry, the file " + filename + " does not exist."
    print(msg)

Sorry, the file alice.txt does not exist.


#### -_Analyzing Text_
The __split()__ method seperates a string into parts wherever it finds a space and stores all the parts of the string in a list. The result is a list of words from the string, although some <br>
punctuation may also appear with some of the words. To count the number of word in _Alice in Wonderland_, we'll use split() on the entire text. Then we'll count the items in the list to get a rough <br>
idea of the number of words in the text:

In [24]:
file_path = "Resources/alice.txt"
try:
    with open(file_path, encoding='utf-8') as f_obj:
        contents = f_obj.read()
except FileNotFoundError:
    msg = "Sorry, the file " + file_path + "doesn't exist."
    print(msg)
else:
    #Count the approximate number of words in the file.
    words = contents.split()
    num_words = len(words)
    print("The file " + file_path + " has about " + str(num_words) + " words.") 

The file Resources/alice.txt has about 29461 words.


#### -_Working with Multiple Files_
Let's add more books to analyze. But before we do, let's move the bulk of this program to a function called __count_words( )__. By doing so, it will be easier to run the analysis for multiple books:

In [25]:
def count_words(filename):
    #Count the approximate number of words in a file
    try:
        with open(filename, encoding = 'utf-8') as f_obj:
            contents = f_obj.read()
    except FileNotFoundError:
        msg = "Sorry, the file " + filename + " does not exist."
        print(msg)
    else:
        #Count approximate number of words in the file.
        words = contents.split()
        num_words = len(words)
        print("The file " + filename + " has about " + str(num_words) + " words.")

file_path = 'Resources/alice.txt'
count_words(file_path)
# filenames = ['alice.txt', 'sidhartha.txt', 'moby_dick.txt', 'little_women.txt']
# for filename in filenames:
#     count_words(filename)

The file Resources/alice.txt has about 29461 words.


output: <br>
The file alice.txt has bout 29461 words. <br>
Sorry, the file siddhartha.txt does not exist. <br>
The file moby_dick.txt has about 215136 words. <br>
The file little_women.txt has about 189079 words. <br>
#### -_Failing Silently_
You don't need to report every exception you catch. You can make a program fail silently, write a __try__ block as usual, but you explicity tell Python to do nothing in the __except__ block. <br>
Python has a __pass__ statement that tells it to do nothing in a block:

In [28]:
def count_words(filename):
    #Count the approximate number of words in a file
    try:
        with open(filename, encoding = 'utf-8') as f_obj:
            contents = f_obj.read()
    except FileNotFoundError:
        pass
    else:
        #Count approximate number of words in the file.
        words = contents.split()
        num_words = len(words)
        print("The file " + filename + " has about " + str(num_words) + " words.")

file_path = 'siddhartha.txt'
count_words(file_path)

The pass statement also acts as a placeholder. It's a reminder that you're choosing to do nothing at a specific point in your program's execution and that you might want to do something there later. <br>
***
### Storing Data 
Many of your programs will ask users to input certain kinds of information. You might allow users to store preferences in a game or provide data for a visualization. Whatever the focus of your <br>
program is, you'll store the information users provide in data structures such as lists and dictionaries. When users close a program, you'll almost always want to save the information they entered.<br>
A simple way to do this involves storing your data using the __json__ module. <br>
<br>
The __json__ module allows you to dump simple Python data structures into a file and load the data from that file the next time the program runs. You can also use json to share data between different<br> 
Python programs. Even better, the JSON data format is not specific to Python, so you can share data you store in the JSON format with people who work in many other programming languages. <br>
<br>
#### -_Using json.dump( ) and json.load( )_ 
Let's write a short program that stores a set of numbers and another program that reads these numbers back into memory. The first program will use __json.dump( )__ to store the set of numbers, and the<br>
second program will use __json.load( )__. <br>
<br>
The __json.dump( )__ function takes two arguments: a piece of data to store and a file object it can use to store the data. Here's how you can use __json.dump( )__ to store list of numbers: